core/executor/src/sandbox.rs

```diff
@@ -175,15 +184,37 @@ pub trait SandboxCapabilities {
 	///
 	/// Returns `Err` if `ptr + len` is out of bounds.
 	fn read_memory(&self, ptr: Pointer<u8>, len: WordSize) -> Result<Vec<u8>>;
+
+	/// Invoke a function in the supervisor environment.
+	///
+	/// This first invokes the dispatch_thunk function, passing in the function index of the
+	/// desired function to call and serialized arguments. The thunk calls the desired function
+	/// with the deserialized arguments, then serializes the result into memory and returns
+	/// reference. The pointer to and length of the result in linear memory is encoded into an i64,
+	/// with the upper 32 bits representing the pointer and the lower 32 bits representing the
+	/// length.
+	///
+	/// # Errors
+	///
+	/// Returns `Err` if the dispatch_thunk function has an incorrect signature or traps during
+	/// execution.
+	fn invoke(
+		&mut self,
+		dispatch_thunk: &Self::SupervisorFuncRef,
+		invoke_args_ptr: Pointer<u8>,
+		invoke_args_len: WordSize,
+		state: u32,
+		func_idx: SupervisorFuncIndex,
+	) -> Result<i64>;
```

Comment:
This is pity that we are exposing this, since I'd be happy to get rid of the dispatch thunk shenanigans, and operate on function references directly.

Reply:
Hmm, that sounds like a pretty big change. Would rather just keep it like this for now.

Reply:
Yep, it is : ) To be clear, I am **not** asking for you to do this, esp. in this PR. I was just iterating on new sandbox v2 designs when I tried to do this.